### DISCUSSION TITLE
Discussion about "The Abstraction and Reasoning Corpus (ARC)"

### DISCUSSION THREAD

**Discussion by antosold** [erodola/DLAI-s2-2022#15] [General]
Hi all! I would like to start here a discussion about "The Abstraction and Reasoning Corpus (ARC)", the wonderful project pointed out by the last exercise in the notebook "Linear_models_and_Pytorch_Datasets". Below what I understood from it and my impressions.

I inspected a bit the goal of that project and basically it is about comparing the AI with the human intelligence by submitting to both some tasks to resolve.
Each task is composed by an algorithm which builds up a pattern that relates an input grid to an output grid (the "pair").
The game is about to extract the right pattern from the 3 pairs proposed to the left (I highligh that that patterns is different for each task)

![image](https://user-images.githubusercontent.com/101824957/158982513-f04d960d-a778-4927-a575-7576037c9a46.png)

and use that to test the human intelligence on the right

![image](https://user-images.githubusercontent.com/101824957/158982786-354d892a-15cf-48ed-bc9b-46cf695810e1.png)

In this task, the correct solution is a 3x3 grid full of light blue squares: the pattern is based on counting occurrences of the colours in the grid an full fill the output one with the color which appears the most.

I report here the exact description found in the [paper here](https://arxiv.org/pdf/1911.01547.pdf), maybe it could be more clear :)

_When solving an evaluation task, a test-taker has access to the training examples for the task (both the input and output grids), as well as the input grid of the test examples for the task. The test-taker must construct on its own the output grid corresponding to the input grid of each test example. “Constructing the output grid” is done entirely from scratch, meaning that the test-taker must decide what the height and width of the output grid should be, what symbols it should place on the grid, and where. The task is successfully solved if the testtaker can produce the exact correct answer on all test examples for the task (binary measure of success)._

I encourage you to read on the paper linked before at page 48 the prior used to build up the algorithm: it mentions **Object cohesion**, **Object persistence** and **Object influence via contact**, which I came upon in the tests I took while playing. In fact, the correct output grid often had to be build up by symmetrizing, let the colored squares be in contact with each other somehow or just counting the occurrences.

Coming to the question asked in the notebook:

- What would your Dataset return for this task?
- What would each batch contain?

In my opinion, the Dataset is simply a dictionary of tasks (in other words, each json file for the training set) so it returns a task which is a set of pairs (input-output grids). The AI algorithm runs over them trying to intercept the right pattern to build. Then, the evaluation set is prompted to the user in order to verify the correctness of the algo.
The batch is composed by the 3 demonstration tasks proposed to the user.

I'm not 100% sure about that but those are my thoughts. I'm so glad if you partecipate to this discussion in order to share our impressions.
Plus, I propose you another question: what do you think could be the method used by the AI to minimize the error and predict the correct grid?

Thank you all!

**Comment by erodola** [comment]
Thank you for bringing this up!
As a matter of fact, abstraction and reasoning are still very open aspects in AI, that the ARC initiative only partly addressed (if at all).
You are asking valid and stimulating questions, and it would be great to hear some input from your fellow students!

**Comment by alesspalma** [comment]
Here are my answers:
- **What would your Dataset return for this task?** Considering the data/training folder, I will create a Dataset class wich returns as data x for the i-th index all the input grids of the i-th task (both "train" and "test" inputs), and as ground truth y all its output grids. It could be the train set for my model, meanwhile for the test set I will create the same Dataset but with data in the data/evaluation folder.

- **What would each batch contain?** Up to me the most reasonable choice is to make each batch contain a single task, in order to make the model learn on that single task and then pass to the subsequent, with the hope of learning how to generalize well. It means that actually my Dataset is already batched (batch_size of Dataloader should be 1 then), but if we want to modify this behaviour we could just flatten the Dataset x and y and mix up things.

 - **What do you think could be the method used by the AI to minimize the error and predict the correct grid?** This is really a hard question up to me, honestly I am not sure that neural networks would be the best way to go for such an inference/reasoning problem (I also think we have too few data to train a neural network), but if I had to I would treat each output pixel as a classification problem, where classes are the 10 possible colors of the single pixel. So the loss would be something like the average of the Cross-Entropy losses of each output neuron. Probably there exist better/simpler algorithms that could generalize better (maybe reinforcement learning?).

I would also like to give back to you 2 questions:
- What's your answer to your last question?
- I didn't catch if you are going to drop the "test" image pair (input:output) in each train task, because you said that the batch would be composed by 3 images, meanwhile if we consider the 3 "train" pairs + 1 "test" pair (on average) in a single .json task, then that would be 4. Instead, if you are using the "test" pairs in the task of the data/training folder as your test set, then you are not using the data/evaluation folder (of course is not wrong, is just a choice, I am just interested in better understanding what's your choice).

Thanks for the interesting discussion :)
